ir/src/ir/value.rs

```diff
@@ -1,15 +1,28 @@
+use crate::NodeIndex;
+
 use super::*;
 
-/// Represents a scalar value in the [AlgebraicGraph]
+/// Represents a scalar value in the [MIR]
 ///
 /// Values are either constant, or evaluated at runtime using the context
 /// provided to an AirScript program (i.e. random values, public inputs, etc.).
-#[derive(Debug, Eq, PartialEq, Copy, Clone)]
-pub enum Value {
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum MirValue {
```

Comment:
Our current design for the IR is closer to a sea-of-nodes than the more common CFG representation. The main difference with a classic sea-of-nodes implementation is that we do preserve ordering in the function/block bodies. This also means that our SSA values can be extrapolated deterministically when pretty printing rather than being stored in the nodes directly. We currently do this by traversing block bodies in order, but could also be done with post-order traversal of the dependency graph.
We have a few questions/remarks about your proposed structure:
1) What would be the advantage of using an intrusive linked list here?
2) Switch from sea-of-nodes to CFG: might delay work on passes and need a rewrite of the lowering AST->MIR. The 2 models are valid for an IR, with varying advantages and drawbacks.
3) We agree that the node types should be more properly separated, we'll tend towards your proposed design on this aspect over time, incorporating elements as we need them for this implementation.
4) Keeping the regionalised owner attributes would help simplify passes greatly. We'll incorporate that into our design.
Overall we think we can benefit from the proposed changes without necessarily changing the core design of the sea-of-nodes system we already have. This also minimizes time spent on redoing existing work.

Reply:
> Our current design for the IR is closer to a sea-of-nodes than the more common CFG representation.

I should note that this is true of the representation described above as well - more precisely, it is a simplified representation of a Regionalized Value-State Dependence Graph (RVSDG), where the connections between nodes are entirely dictated by the value dependency graph. The exception to this, if you can call it that, are "SSA regions", i.e. regions with multiple blocks, where values are expected to adhere to the SSA dominance property - but you can also have regions where all operations go in a single block with no implicit scheduling order, beyond that dictated by the value dependency graph. It is basically up to the semantics of each operation to dictate what kind of regions it has. MLIR, which uses a similar representation, names these two kinds of regions "SSACFG" regions and "graph" regions.

> This also means that our SSA values can be extrapolated deterministically when pretty printing rather than being stored in the nodes directly.

The use of `ValueId` above is really one of debugging convenience, you can absolutely skip it, and assign integral identifiers during pretty printing. 

The tradeoff of that, is that it means that if you are debugging an issue with the compiler, dumping a vector of values doesn't give you a way to correlate those values to what you see in the pretty printed output (and from that, to the source code), all you get are opaque values (and some associated information like types, but that is commonly not enough to know precisely which value you are looking at, or how it was derived). It is convenient to be able to correlate the output from something like `dbg!(value)` with the pretty printed output and source code of the program/function/region, so that you can better understand what is happening. In any case, you could certainly omit it, I just find in practice that it is worth assigning a debugging-friendly canonical identifier for uses like that (one can also just perform this only when under `#[cfg(debug_assertions)]` or something, plenty of different ways to approach things of this nature).

> What would be the advantage of using an intrusive linked list here?

The biggest benefits are:

* You don't need to allocate any memory for containers that a node might be a member of, e.g., the list of operations in a block, a list of users of some symbol, all can be maintained by links between nodes directly. 
* It lets you manipulate nodes in the graph independently of their parents. 
* It permits a node to be present in multiple lists at once. For example, if you wanted to track symbol usage (where operations can be symbols), then an operation might have an intrusive link both for its parent block _and_ the use list of the symbol it uses (if it uses one).
* Many common transformations/rewrites rely on: inserting nodes at the front or back of some list of nodes, inserting a node relative to some other node; moving nodes around relative to their current position, or to the position of another node;  appending a list of nodes to another list, splicing a list of nodes into the middle of another list; splitting a list of nodes at arbitrary positions; and "stealing" a list of nodes from its current owner, leaving an empty list in its place. The intrusive linked-list is very efficient at all of these operations, in particular it is a constant-time operation to construct a cursor to a node in its containing list, so that you can then perform some operation on the list relative to that cursor (e.g. removing the node, splitting the list at that node, splicing another list at that node, inserting a new node). Vectors are great for cache locality, iteration, indexed access, and pushing to the end (assuming it doesn't need to reallocate the backing storage in the process) - but they are much less efficient at all of the other operations I mentioned above. 

Basically, it is an excellent data structure for a compiler IR which relies on an implicit graph, rather than an explicit one. The primary downside of linked lists is the potential for poor cache locality - but that is mitigated significantly by allocating nodes from an arena. I didn't go that far with my example code, but in practice that is usually what I do. Most traversals of an IR are jumping between different node types (regions -> blocks -> operations -> values -> operations -> blocks -> values -> ...), so cache locality is inherently an issue even when using vectors for node storage. Vectors are of course the primary alternative, using them as an arena, and using handles (indexes in some vector of a specific node type) for node-to-node references - implementing things like lists of regions, blocks, operations, etc., as vectors of handles. It's not a _bad_ alternative by any means, perhaps even better in some regards, but it does make some things more awkward in Rust, particularly mutation.


> Switch from sea-of-nodes to CFG: might delay work on passes and need a rewrite of the lowering AST->MIR. The 2 models are valid for an IR, with varying advantages and drawbacks.

As an aside: I want to push back on characterizing this as a CFG representation - it is an RVSDG, which is a form of dataflow graph. To the extent that there is any control flow represented, it is by convention only (i.e. the fact that operations are ordered in a block is incidental). What actually dictates whether a given region must _also_ represent a CFG, is whether or not you implement analyses/transformations over it that depend upon the SSA dominance property. Since AirScript doesn't have any real notion of control flow, only data flow, I think it is fair to say that we don't have any particular reason to require this of our IR, regardless of representation.

In any case, I certainly agree that switching representations is probably not worth it at this point, since I don't see any reason to assume that the current approach is not going to work out. My primary purpose in bringing it up and elaborating on it is two-fold:

1. The current sea-of-nodes representation is fine, aside from nits like splitting up the representation of values and operations, but I have two concerns (not unresolvable ones, but want to note them), and I wanted to frame it in comparison to the alternative approach that I probably would have taken if I was the one doing the work:
  * Uniquing nodes, and the interaction with program transformations. It may be that this isn't actually an issue in practice, but we'll want to keep an eye on this, and test thoroughly all of the potential edge cases.
  * The ability for less senior engineers to reason about this representation and work on the AirScript compiler. More thorough documentation once things are firmed up will be essential, particularly any assumptions made that have implications for correctness of the translation from the AST. For example, at what points are bindings resolved to SSA values, and does that occur before/after uniquing (presumably all bindings are resolved before uniquing, as let-shadowing is permitted in AirScript, but all of those kind of details should be outlined).
2. I had outlined the RVSDG design in our previous discussions elsewhere, (in #358 IIRC), but not elaborated on what that would actually look like in practice. I wanted to do so here in order to compare/contrast, and provide you with a concrete idea of what I'm talking about if I refer to things not present in this specific IR.

To be 100% clear, it was more to provide a background for discussion about this IR, rather than to dictate a new design.

> We agree that the node types should be more properly separated, we'll tend towards your proposed design on this aspect over time, incorporating elements as we need them for this implementation.

👍

>  Keeping the regionalised owner attributes would help simplify passes greatly. We'll incorporate that into our design.

I'm assuming it will be necessary in order to do per-region uniquing. However, after mulling it over further, perhaps it is not necessary to do per-region uniquing? I would like to better understand the implications of global uniquing, and I think that requires a broader set of tests and implemented functionality to see how it all interacts. It may very well be that there are no adverse effects because of how simple AirScript is in terms of primitive operations.

> Overall we think we can benefit from the proposed changes without necessarily changing the core design of the sea-of-nodes system we already have. This also minimizes time spent on redoing existing work.

Sure, I think that's fine. At the very least, you have an example of an alternative approach if you find you need to elaborate on the current representation to address any issues you run into. Again, the idea was more to compare/contrast and provide an explanation of what I'm talking about, rather than dictate how you implement the IR.

Reply:
> I should note that this is true of the representation described above as well - more precisely, it is a simplified representation of a Regionalized Value-State Dependence Graph (RVSDG), where the connections between nodes are entirely dictated by the value dependency graph

I see, I think we misinterpreted your proposed implementation here, you can safely discard any mention of a CFG in our previous comment.
After thinking more about it, your design does seem to present another advantage in the case of dead code elimination. Since nodes are referenced by pointer (or `Rc`) and not their NodeIndex in the Vec of all nodes, this would enable automatic cleaning of unreferenced nodes without worrying about having to shift subsequent NodeIndices in order to compactify the graph. Using reference counting makes this a lot easier to manage than doing it manually

> The use of ValueId above is really one of debugging convenience, you can absolutely skip it, and assign integral identifiers during pretty printing.

Noted, this is a very good point I hadn't thought about, we'll add tracking of ssa values to the nodes

> The biggest benefits are: ... 

I see what you mean, thanks for the detailed explanation. We'll probably not implement this immediately, it would probably be worth measuring real performance before optimizing this part. We'll keep that structure in mind as a potential future optimization.

> As an aside: I want to push back on characterizing this as a CFG representation - it is an RVSDG, which is a form of dataflow graph

Noted, that is mostly due to our misunderstanding of your proposed design. After thinking about it longer, I am under the impression that the 2 designs are functionally equivalent.

> To be 100% clear, it was more to provide a background for discussion about this IR, rather than to dictate a new design.

No problem! And thank you for this detailed feedback, it clears a few points I hadn't understood properly.

> I'm assuming it will be necessary in order to do per-region uniquing. However, after mulling it over further, perhaps it is not necessary to do per-region uniquing? I would like to better understand the implications of global uniquing, and I think that requires a broader set of tests and implemented functionality to see how it all interacts. It may very well be that there are no adverse effects because of how simple AirScript is in terms of primitive operations.

That was our impression aswell, it seemed a bit irrelevant after the inlining/unrolling passes since they remove all function "scopes" and only keeps the root block. Note that we havent fully thought through the implications for conditional's and loop's blocks.
Our current understanding is that keeping it regionalized makes sense if we want to have pre-inlining/unrolling optimizations, since after inlining/unrolling, those would all be sharing the same scope anyway.

We'll implement regionalisation for now, but might try to remove that requirement in the future if at all possible.

> At the very least, you have an example of an alternative approach if you find you need to elaborate on the current representation to address any issues you run into. Again, the idea was more to compare/contrast and provide an explanation of what I'm talking about, rather than dictate how you implement the IR.

Thanks again for this!